src/catalog/catalog.cpp

```diff
@@ -832,28 +868,47 @@ const FunctionData Catalog::GetFunction(
     throw CatalogException("Failed to find function " + name);
   }
 
-  // If the language isn't 'internal', crap out ... for now ...
+  // If the language isn't 'internal' or 'plpgsql', crap out ... for now ...
   auto lang_catalog_obj = proc_catalog_obj->GetLanguage();
   if (lang_catalog_obj == nullptr ||
-      lang_catalog_obj->GetName() != "internal") {
+      (lang_catalog_obj->GetName() != "internal" &&
+       lang_catalog_obj->GetName() != "plpgsql")) {
     txn_manager.AbortTransaction(txn);
     throw CatalogException(
-        "Peloton currently only supports internal functions. Function " + name +
-        " has language '" + lang_catalog_obj->GetName() + "'");
+        "Peloton currently only supports internal functions and plpgsql UDFs. \
+        Function " +
+        name + " has language '" + lang_catalog_obj->GetName() + "'");
   }
 
-  // If the function is "internal", perform the lookup in our built-in
-  // functions map (i.e., function::BuiltInFunctions) to get the function
   FunctionData result;
   result.argument_types_ = argument_types;
   result.func_name_ = proc_catalog_obj->GetSrc();
   result.return_type_ = proc_catalog_obj->GetRetType();
-  result.func_ = function::BuiltInFunctions::GetFuncByName(result.func_name_);
-
-  if (result.func_.impl == nullptr) {
-    txn_manager.AbortTransaction(txn);
-    throw CatalogException("Function " + name +
-                           " is internal, but doesn't have a function address");
+  if (lang_catalog_obj->GetName() == "internal") {
+    // If the function is "internal", perform the lookup in our built-in
+    // functions map (i.e., function::BuiltInFunctions) to get the function
+    result.func_ = function::BuiltInFunctions::GetFuncByName(result.func_name_);
+    result.isUDF_ = false;
+    if (result.func_.impl == nullptr) {
+      txn_manager.AbortTransaction(txn);
+      throw CatalogException(
+          "Function " + name +
+          " is internal, but doesn't have a function address");
+    }
+  } else if (lang_catalog_obj->GetName() == "plpgsql") {
+    // If the function is a "plpgsql" udf, perform the lookup in the plpgsql
+    // functions map (i.e., function::PlpgsqlFunctions) to get the function
+    // code_context
+    result.func_context_ = function::PlpgsqlFunctions::GetFuncContextByOid(
+        proc_catalog_obj->GetOid());
+    result.isUDF_ = true;
+
+    if (result.func_context_->GetUDF() == nullptr) {
```

Comment:
Is this ever possible?

Reply:
Technically, if there are any exceptions raised while parsing the UDF definition, this code will be entered.